File: agent/commands-generator/src/main/java/org/matilda/commands/utils/Package.java
package org.matilda.commands.utils;

import java.nio.file.FileSystems;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

public class Package {
    private final List<String> mParts;

    public Package(List<String> parts) {
        mParts = parts.stream()
                .filter(part -> !part.isEmpty())
                .collect(Collectors.toList());
    }

    public Package(String... parts) {
        this(List.of(parts));
    }

    public Package subpackage(String child) {
        return subpackage(fromString(child));
    }

    private Package subpackage(Package child) {
        return joinPackages(this, child);
    }

    public static Package fromString(String packageName) {
        return new Package(packageName.split("\\."));
    }

    public static Package joinPackages(Package... packages) {
        return new Package(Arrays.stream(packages)
                .flatMap(p -> p.mParts.stream())
                .collect(Collectors.toList()));
    }

    public String getPackageName() {
        return String.join(".", mParts);
    }

    public String getLastPart() {
        return mParts.get(mParts.size() - 1);
    }

    public Package withoutLastPart() {
        return new Package(mParts.subList(0, mParts.size() - 1));
    }

    public Package removeCommonPrefixFrom(Package other) {
        int equalParts = 0;
        while (equalParts < mParts.size() && equalParts < other.mParts.size()) {
            if (!mParts.get(equalParts).equals(other.mParts.get(equalParts))) {
                break;
            }
            equalParts++;
        }
        return new Package(mParts.subList(equalParts, mParts.size()));
    }

    public List<String> getParts() {
        return mParts;
    }

    public String toPath() {
        return String.join(FileSystems.getDefault().getSeparator(), mParts);
    }
}

File: agent/commands-generator/src/main/java/org/matilda/commands/utils/Package.kt
package org.matilda.commands.utils

import java.io.File

data class Package(val parts: List<String>) {
    constructor(vararg parts: String) : this(listOf(*parts))

    fun subpackage(child: String) = subpackage(fromString(child))

    fun subpackage(child: Package) = joinPackages(this, child)

    val packageName: String
        get() = parts.joinToString(".")
    val lastPart: String
        get() = parts.last()

    fun withoutLastPart() = Package(parts.dropLast(1))

    fun removeCommonPrefixFrom(other: Package): Package {
        var equalParts = 0
        while (equalParts < parts.size && equalParts < other.parts.size) {
            if (parts[equalParts] != other.parts[equalParts]) {
                break
            }
            equalParts++
        }
        return Package(parts.subList(equalParts, parts.size))
    }

    fun toPath() = parts.joinToString(File.separator)

    companion object {
        fun fromString(packageName: String) = Package(packageName.split("\\.".toRegex()))

        fun joinPackages(vararg packages: Package) = Package(packages.flatMap(Package::parts))
        fun fromPath(path: String): Package = Package(path.split(File.separator).filter { it.isNotEmpty() })
    }
}
